File: cli/cmd/kaioken/ext.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"strings"

	"kaioken/internal/ext"
)

const extUsage = `usage: kaioken ext <subcommand>

  install <owner/repo[@version]>   install an extension from its GitHub release
  dev <path>                       install from a local directory (author dev loop)
  validate [path]                  lint an extension working tree before publishing
  remove <id>                      uninstall an extension
  list                             show installed extensions (the default)
  update [id]                      update every extension (or one) to the latest release
  search [term]                    browse the community registry
  enable <id> | disable <id>       toggle an installed extension
  trust <id>                       allow an mcp/wasm extension to run (-force skips the prompt)
  untrust <id>                     revoke that permission and stop it
  tools <id>                       list an extension's tools (-force re-queries the plugin)

browse the catalog on the web: https://extensions.kaioken.dev
registry override: set ext_registry in ~/.kaioken/config.yaml (any URL serving
the community-extensions.json shape, e.g. a company-internal index)
`

// cmdExt manages community extensions: declarative packages installed from
// GitHub releases into ~/.kaioken/extensions. See internal/ext.
func cmdExt(ctx context.Context, f flags) error {
	sub, arg := "", ""
	if len(f.positionals) > 0 {
		sub = strings.ToLower(f.positionals[0])
	}
	if len(f.positionals) > 1 {
		arg = f.positionals[1]
	}

	switch sub {
	case "", "list", "ls":
		return extList()
	case "install", "add":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext install <owner/repo[@version]>")
		}
		return extInstall(ctx, arg)
	case "dev":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext dev <path>")
		}
		return extDev(arg)
	case "validate", "lint":
		if arg == "" {
			arg = "."
		}
		return extValidate(arg)
	case "remove", "uninstall", "rm":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext remove <id>")
		}
		if err := ext.Remove(arg); err != nil {
			return err
		}
		fmt.Printf("  ✓ removed %s\n", arg)
		return nil
	case "update":
		return extUpdate(ctx, arg)
	case "search":
		return extSearch(ctx, arg)
	case "enable", "disable":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext %s <id>", sub)
		}
		if err := ext.SetEnabled(arg, sub == "enable"); err != nil {
			return err
		}
		fmt.Printf("  ✓ %s %sd\n", arg, sub)
		return nil
	case "trust":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext trust <id>")
		}
		return extTrust(ctx, arg, f.force)
	case "untrust":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext untrust <id>")
		}
		if err := ext.Untrust(arg); err != nil {
			return err
		}
		fmt.Printf("  ✓ %s untrusted — its server will not run again until you re-trust it\n", arg)
		return nil
	case "tools":
		if arg == "" {
			return fmt.Errorf("usage: kaioken ext tools <id>")
		}
		return extTools(ctx, arg, f.force)
	default:
		return fmt.Errorf("unknown ext subcommand %q\n\n%s", sub, extUsage)
	}
}

func extList() error {
	lock, err := ext.LoadLock()
	if err != nil {
		return err
	}
	if len(lock.Extensions) == 0 {
		fmt.Println("no extensions installed")
		fmt.Println("  kaioken ext search           browse the community registry")
		fmt.Println("  kaioken ext install o/r      install straight from a GitHub repo")
		return nil
	}
	for _, e := range lock.Extensions {
		state := "enabled"
		if !e.Enabled {
			state = "disabled"
		}
		trust := "-"
		if man, _, err := ext.InstalledManifest(e.ID); err == nil && ext.Executable(man.Type) {
			if e.Trusted() {
				trust = "trusted"
			} else {
				trust = "UNTRUSTED"
			}
		}
		fmt.Printf("  %-32s %-10s %-28s %-9s %s\n", e.ID, e.Version, e.Repo, state, trust)
	}
	fmt.Printf("%d extension(s) in %s\n", len(lock.Extensions), ext.Root())
	return nil
}

func extInstall(ctx context.Context, source string) error {
	fmt.Printf("installing %s …\n", source)
	res, err := ext.Install(ctx, source)
	if err != nil {
		return err
	}
	for _, w := range res.Warnings {
		fmt.Printf("  ! %s\n", w)
	}
	fmt.Printf("  ✓ %s %s  (%s, release %s)\n", res.Entry.ID, res.Entry.Version, res.Entry.Repo, res.Entry.Tag)
	for _, s := range res.Skills {
		fmt.Printf("    + skill %s — %s\n", s.Name, s.Description)
	}
	if len(res.Skills) == 0 {
		fmt.Println("    (no skills contributed)")
	}
	if res.NeedsTrust {
		switch res.Manifest.Type {
		case ext.TypeWasm:
			fmt.Println("\n  ! this is a wasm extension: it ships sandboxed plugin code")
			fmt.Printf("    module: %s   permissions: %s\n", res.Manifest.Wasm.Entry, permissionLine(res.Manifest))
		default:
			fmt.Println("\n  ! this is an mcp extension: it declares a server that would run UNSANDBOXED")
			fmt.Printf("    command: %s\n", commandLine(res.Manifest))
		}
		fmt.Printf("    it stays inert until you run `kaioken ext trust %s`\n", res.Entry.ID)
	}
	return nil
}

// commandLine renders an mcp manifest's launch invocation for display — the
// exact thing trusting would allow to run.
func commandLine(man *ext.Manifest) string {
	if man == nil || man.MCP == nil {
		return "(none)"
	}
	parts := append([]string{man.MCP.Command}, man.MCP.Args...)
	line := strings.Join(parts, " ")
	if len(man.MCP.Env) > 0 {
		var envs []string
		for k, v := range man.MCP.Env {
			envs = append(envs, k+"="+v)
		}
		line += "   (env: " + strings.Join(envs, " ") + ")"
	}
	return line
}

// permissionLine renders a wasm manifest's capability grants for display.
func permissionLine(man *ext.Manifest) string {
	if len(man.Permissions) == 0 {
		return "none — fully isolated"
	}
	return strings.Join(man.Permissions, ", ")
}

// extTrust shows exactly what would run, asks, and only then grants trust —
// which runs the plugin once to record its tool catalog.
func extTrust(ctx context.Context, id string, force bool) error {
	man, entry, err := ext.InstalledManifest(id)
	if err != nil {
		return err
	}
	if !ext.Executable(man.Type) {
		return fmt.Errorf("%s is a declarative extension — it runs no code and needs no trust", id)
	}
	if entry.Trusted() {
		fmt.Printf("%s %s is already trusted\n", id, entry.Version)
		return nil
	}

	switch man.Type {
	case ext.TypeWasm:
		fmt.Printf("%s %s wants to run a SANDBOXED wasm plugin:\n", id, entry.Version)
		fmt.Printf("\n    module:      %s\n    permissions: %s\n\n", man.Wasm.Entry, permissionLine(man))
		fmt.Println("The sandbox gives it no network, no environment, and no filesystem")
		fmt.Println("beyond the permissions listed (fs:read:workspace mounts your repo")
		fmt.Println("read-only). Memory is capped. Trust applies to this exact version;")
		fmt.Println("updates revoke it.")
	default:
		fmt.Printf("%s %s wants to run an MCP server on your machine, UNSANDBOXED:\n", id, entry.Version)
		fmt.Printf("\n    %s\n\n", commandLine(man))
		fmt.Println("Kaioken cannot restrict what this process does. Trust it only if you")
		fmt.Println("trust its author. Trust applies to this exact version; updates revoke it.")
	}
	if !force {
		fmt.Print("\nTrust and run it now? [y/N]: ")
		line, _ := bufio.NewReader(os.Stdin).ReadString('\n')
		switch strings.ToLower(strings.TrimSpace(line)) {
		case "y", "yes":
		default:
			fmt.Println("not trusted — nothing was run")
			return nil
		}
	}

	tools, err := ext.Trust(ctx, id)
	if err != nil {
		return err
	}
	fmt.Printf("  ✓ %s %s trusted — %d tool(s) now available to the agent (each call still needs approval)\n",
		id, entry.Version, len(tools))
	printExtTools(tools)
	return nil
}

func extTools(ctx context.Context, id string, refresh bool) error {
	var tools []ext.Tool
	var err error
	if refresh {
		tools, err = ext.RefreshTools(ctx, id)
	} else {
		tools, err = ext.CachedTools(id)
	}
	if err != nil {
		return err
	}
	if len(tools) == 0 {
		fmt.Println("no tools recorded — trust the extension first, or -force to re-query the plugin")
		return nil
	}
	printExtTools(tools)
	return nil
}

func printExtTools(tools []ext.Tool) {
	for _, t := range tools {
		fmt.Printf("    %-40s %s\n", t.FullName, t.Description)
	}
}

func extUpdate(ctx context.Context, id string) error {
	var results []ext.UpdateResult
	var err error
	if id != "" {
		results, err = ext.Update(ctx, id)
	} else {
		results, err = ext.Update(ctx)
	}
	if err != nil {
		return err
	}
	if len(results) == 0 {
		fmt.Println("no extensions installed")
		return nil
	}
	for _, r := range results {
		switch {
		case r.Err != nil:
			fmt.Printf("  ✗ %s: %v\n", r.ID, r.Err)
		case r.Local:
			fmt.Printf("  · %s %s (local dev install — refresh with `kaioken ext dev`)\n", r.ID, r.From)
		case r.Updated:
			fmt.Printf("  ✓ %s %s → %s\n", r.ID, r.From, r.To)
		default:
			fmt.Printf("  · %s %s is current\n", r.ID, r.From)
		}
	}
	return nil
}

// extDev installs from a local working tree — the author's dev loop.
func extDev(path string) error {
	res, err := ext.InstallDev(path)
	if err != nil {
		return err
	}
	for _, w := range res.Warnings {
		fmt.Printf("  ! %s\n", w)
	}
	fmt.Printf("  ✓ %s %s installed from %s (dev)\n", res.Entry.ID, res.Entry.Version, path)
	for _, s := range res.Skills {
		fmt.Printf("    + skill %s — %s\n", s.Name, s.Description)
	}
	if res.NeedsTrust {
		fmt.Printf("    executable extension — trust it with `kaioken ext trust %s`\n", res.Entry.ID)
	}
	fmt.Println("  re-run `kaioken ext dev` after editing the source to refresh")
	return nil
}

// extValidate lints an extension working tree — the same check the template
// repository's CI runs before a release.
func extValidate(path string) error {
	rep, err := ext.ValidateDir(path)
	if err != nil {
		return err
	}
	man := rep.Manifest
	typ := man.Type
	if typ == "" {
		typ = ext.TypeDeclarative
	}
	fmt.Printf("  %s %s  (%s)\n", man.ID, man.Version, typ)
	if len(man.Permissions) > 0 {
		fmt.Printf("  permissions: %s\n", strings.Join(man.Permissions, ", "))
	}
	for _, s := range rep.Skills {
		fmt.Printf("    + skill %s — %s\n", s.Name, s.Description)
	}
	for _, w := range rep.Warnings {
		fmt.Printf("  ! %s\n", w)
	}
	if len(rep.Warnings) == 0 {
		fmt.Println("  ✓ valid — ready to publish")
	} else {
		fmt.Printf("  valid with %d warning(s)\n", len(rep.Warnings))
	}
	return nil
}

func extSearch(ctx context.Context, term string) error {
	entries, err := ext.Registry(ctx, false)
	if err != nil {
		return fmt.Errorf("%w\ndirect install still works: kaioken ext install <owner/repo>", err)
	}
	hits := ext.SearchRegistry(entries, term)
	if len(hits) == 0 {
		fmt.Println("no matching extensions in the registry")
		return nil
	}
	for _, e := range hits {
		fmt.Printf("  %-32s %-28s [%s] %s\n", e.ID, e.Repo, e.TierLabel(), e.Description)
	}
	fmt.Printf("%d result(s) — install with `kaioken ext install <owner/repo>`\n", len(hits))
	fmt.Println("browse the full catalog with READMEs and trust details at https://extensions.kaioken.dev")
	return nil
}

File: kaioken_v2/apps/cli/src/commands/ext.ts
import {
	callMcpTool,
	contributedSkills,
	fetchRegistry,
	findInstalled,
	installExtension,
	isTrusted,
	listInstalled,
	listMcpTools,
	loadLock,
	removeExtension,
	RegistryUnpublished,
	runWasmCommand,
	searchRegistry,
	setEnabled,
	trustExtension,
	updateExtensions,
	type Installed,
} from "@kaioken/ext";
import { VERSION } from "../version.js";
import type { Flags } from "../main.js";

/**
 * `kaioken ext <subcommand>` — community extensions.
 *
 * Three tiers, and the difference between them is the only thing a user really
 * has to understand. A `declarative` extension ships documents: skills in the
 * same format the repository's own use, loaded into the same catalog, and
 * nothing it ships is ever executed. An `mcp` extension declares a server
 * process; a `wasm` extension ships a sandboxed module. Both of those are code
 * from somebody else, so both install inert and stay inert until the exact
 * installed version is trusted by name.
 *
 * That is why `install` and `trust` are separate commands rather than one with
 * a prompt: the gap between them is where a person gets to look at what they
 * just fetched.
 */
export async function runExt(flags: Flags): Promise<number> {
	const [subcommand = "list", ...rest] = flags.positional;

	switch (subcommand.toLowerCase()) {
		case "list":
			return list(flags);
		case "install":
		case "add":
			return install(rest[0] ?? "", rest[1], flags);
		case "remove":
		case "uninstall":
			return remove(rest[0] ?? "", flags);
		case "update":
			return update(rest, flags);
		case "enable":
			return toggle(rest[0] ?? "", true, flags);
		case "disable":
			return toggle(rest[0] ?? "", false, flags);
		case "trust":
			return trust(rest[0] ?? "", rest[1] !== "off", flags);
		case "search":
			return search(rest.join(" "), flags);
		case "tools":
			return tools(rest[0] ?? "", flags);
		case "skills":
			return skills(flags);
		case "run":
			return run(rest[0] ?? "", rest[1] ?? "", rest.slice(2).join(" "), flags);
		default:
			process.stderr.write(
				`kaioken ext: unknown subcommand "${subcommand}"\n` +
					"  list · install · remove · update · enable · disable · trust · search · tools · skills · run\n",
			);
			return 1;
	}
}

async function list(flags: Flags): Promise<number> {
	const installed = await listInstalled();
	if (flags.json) {
		process.stdout.write(`${JSON.stringify(installed, null, 2)}\n`);
		return 0;
	}
	if (installed.length === 0) {
		process.stdout.write("no extensions installed — `kaioken ext search <term>` finds some\n");
		return 0;
	}
	for (const entry of installed) {
		process.stdout.write(`${entry.id}@${entry.version} — ${entry.manifest.name} [${entry.manifest.type}]\n`);
		process.stdout.write(`  ${describeState(entry)}\n`);
	}
	return 0;
}

async function install(source: string, tag: string | undefined, flags: Flags): Promise<number> {
	if (!source) {
		process.stderr.write("kaioken ext install: needs a directory, owner/repo, or a tarball URL\n");
		return 1;
	}
	try {
		const result = await installExtension({
			source,
			hostVersion: VERSION,
			...(tag ? { tag } : {}),
		});
		if (flags.json) {
			process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
			return 0;
		}
		const { entry } = result;
		process.stdout.write(
			`${result.upgraded ? "updated" : "installed"} ${entry.id}@${entry.version} — ${entry.manifest.name}\n`,
		);
		process.stdout.write(`  ${entry.dir}\n`);
		if (entry.manifest.type === "declarative") {
			process.stdout.write("  contributes documents only — nothing it ships is executed\n");
		} else {
			// Said at install time, when the decision is fresh, rather than at
			// first use, when the user has forgotten what they installed.
			// "an mcp" but "a wasm": the article follows the sound, and only two
			// types ever reach this branch.
			const article = entry.manifest.type === "mcp" ? "an" : "a";
			process.stdout.write(
				`  this is ${article} ${entry.manifest.type} extension: it ships code and is INERT until you run\n` +
					`  \`kaioken ext trust ${entry.id}\`\n`,
			);
			if (entry.manifest.permissions?.length) {
				process.stdout.write(`  it asks for: ${entry.manifest.permissions.join(", ")}\n`);
			}
		}
		return 0;
	} catch (error) {
		process.stderr.write(`kaioken ext install: ${error instanceof Error ? error.message : String(error)}\n`);
		return 1;
	}
}

async function remove(id: string, flags: Flags): Promise<number> {
	if (!id) {
		process.stderr.write("kaioken ext remove: needs an extension id\n");
		return 1;
	}
	const removed = await removeExtension(id);
	if (!flags.json) {
		process.stdout.write(removed ? `removed ${id}\n` : `${id} is not installed\n`);
	}
	return removed ? 0 : 1;
}

async function update(ids: readonly string[], flags: Flags): Promise<number> {
	const results = await updateExtensions(ids, { hostVersion: VERSION });
	if (flags.json) {
		process.stdout.write(`${JSON.stringify(results, null, 2)}\n`);
		return 0;
	}
	if (results.length === 0) {
		process.stdout.write("nothing to update\n");
		return 0;
	}
	for (const result of results) {
		process.stdout.write(
			result.updated
				? `${result.id}: ${result.from} → ${result.to}\n`
				: `${result.id}: ${result.reason ?? "unchanged"}\n`,
		);
	}
	return 0;
}

async function toggle(id: string, enabled: boolean, flags: Flags): Promise<number> {
	if (!id) {
		process.stderr.write("kaioken ext: needs an extension id\n");
		return 1;
	}
	const found = await setEnabled(id, enabled);
	if (!flags.json) {
		process.stdout.write(found ? `${id} ${enabled ? "enabled" : "disabled"}\n` : `${id} is not installed\n`);
	}
	return found ? 0 : 1;
}

async function trust(id: string, trusted: boolean, flags: Flags): Promise<number> {
	if (!id) {
		process.stderr.write("kaioken ext trust: needs an extension id — `kaioken ext trust <id> [off]`\n");
		return 1;
	}
	const entry = await trustExtension(id, trusted);
	if (!entry) {
		process.stderr.write(`kaioken ext trust: ${id} is not installed\n`);
		return 1;
	}
	if (flags.json) {
		process.stdout.write(`${JSON.stringify(entry, null, 2)}\n`);
		return 0;
	}
	process.stdout.write(
		trusted
			? `trusted ${entry.id}@${entry.version} to run — the next version will need trusting again\n`
			: `withdrew trust from ${entry.id}\n`,
	);
	return 0;
}

async function search(term: string, flags: Flags): Promise<number> {
	let entries: Awaited<ReturnType<typeof fetchRegistry>>;
	try {
		entries = await fetchRegistry({ force: flags.force });
	} catch (error) {
		// No index published is the ordinary case today, not a fault: reporting
		// it as an error sends the reader hunting for a network problem they do
		// not have, and `ext install <owner/repo>` still works without one.
		if (error instanceof RegistryUnpublished) {
			process.stdout.write(`${error.message}\n`);
			return 0;
		}
		process.stderr.write(`kaioken ext search: ${error instanceof Error ? error.message : String(error)}\n`);
		return 1;
	}
	const hits = searchRegistry(entries, term);

	if (flags.json) {
		process.stdout.write(`${JSON.stringify(hits, null, 2)}\n`);
		return 0;
	}
	if (hits.length === 0) {
		process.stdout.write(term ? `nothing in the registry matches "${term}"\n` : "the registry is empty\n");
		return 0;
	}
	for (const hit of hits.slice(0, 25)) {
		process.stdout.write(`${hit.id} — ${hit.name}${hit.type ? ` [${hit.type}]` : ""}\n`);
		if (hit.description) process.stdout.write(`  ${hit.description}\n`);
		process.stdout.write(`  install: kaioken ext install ${hit.repo}\n`);
	}
	// A listing is not an endorsement, and the trust prompt is what actually
	// decides whether any of this runs.
	process.stdout.write("\nthe registry says an extension exists; it does not vouch for it\n");
	return 0;
}

/** What an executable extension would contribute, asked of the extension itself. */
async function tools(id: string, flags: Flags): Promise<number> {
	const entry = await resolveEntry(id);
	if (!entry) return 1;

	if (entry.manifest.type === "declarative") {
		process.stdout.write(`${entry.id} contributes documents, not tools — \`kaioken ext skills\` lists them\n`);
		return 0;
	}
	if (!isTrusted(entry)) {
		process.stderr.write(
			`kaioken ext tools: ${entry.id}@${entry.version} is not trusted, so nothing was started.\n` +
				`  \`kaioken ext trust ${entry.id}\` after reading what it does\n`,
		);
		return 1;
	}

	if (entry.manifest.type === "wasm") {
		const commands = entry.manifest.commands ?? [];
		if (flags.json) {
			process.stdout.write(`${JSON.stringify(commands, null, 2)}\n`);
			return 0;
		}
		if (commands.length === 0) {
			process.stdout.write(`${entry.id} declares no commands\n`);
			return 0;
		}
		for (const command of commands) {
			process.stdout.write(`${command.name}${command.description ? ` — ${command.description}` : ""}\n`);
		}
		return 0;
	}

	try {
		const list = await listMcpTools(entry);
		if (flags.json) {
			process.stdout.write(`${JSON.stringify(list, null, 2)}\n`);
			return 0;
		}
		for (const tool of list) {
			process.stdout.write(`${tool.name}${tool.description ? ` — ${tool.description}` : ""}\n`);
		}
		if (list.length === 0) process.stdout.write(`${entry.id} exposes no tools\n`);
		return 0;
	} catch (error) {
		process.stderr.write(`kaioken ext tools: ${error instanceof Error ? error.message : String(error)}\n`);
		return 1;
	}
}

async function skills(flags: Flags): Promise<number> {
	const { skills: contributed, problems } = await contributedSkills();
	if (flags.json) {
		process.stdout.write(`${JSON.stringify({ skills: contributed, problems }, null, 2)}\n`);
		return 0;
	}
	for (const skill of contributed) {
		process.stdout.write(`${skill.name} — ${skill.description}\n`);
	}
	if (contributed.length === 0) process.stdout.write("no extension contributes skills\n");
	for (const problem of problems) {
		process.stderr.write(`unusable: ${problem.path} (${problem.extension}) — ${problem.reason}\n`);
	}
	return 0;
}

/** `kaioken ext run <id> <command> [args]` — the `/x` path. */
async function run(id: string, command: string, args: string, flags: Flags): Promise<number> {
	const entry = await resolveEntry(id);
	if (!entry) return 1;
	if (!command) {
		process.stderr.write("kaioken ext run: needs a command — `kaioken ext tools <id>` lists them\n");
		return 1;
	}
	if (entry.manifest.type === "declarative") {
		process.stderr.write(`kaioken ext run: ${entry.id} ships documents, not commands\n`);
		return 1;
	}
	if (!isTrusted(entry)) {
		process.stderr.write(
			`kaioken ext run: ${entry.id}@${entry.version} is not trusted — \`kaioken ext trust ${entry.id}\`\n`,
		);
		return 1;
	}

	try {
		if (entry.manifest.type === "wasm") {
			const result = await runWasmCommand(entry, { command, input: args });
			if (flags.json) {
				process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
				return 0;
			}
			for (const line of result.logs) process.stderr.write(`  ${line}\n`);
			process.stdout.write(`${result.output}\n`);
			return 0;
		}

		// An MCP tool takes structured arguments; a command line gives one
		// string. `input` is the conventional single-field shape, and anything
		// richer is passed as JSON.
		const parsed = parseArguments(args);
		const result = await callMcpTool(entry, command, parsed);
		if (flags.json) {
			process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
			return result.isError ? 1 : 0;
		}
		process.stdout.write(`${result.text}\n`);
		return result.isError ? 1 : 0;
	} catch (error) {
		process.stderr.write(`kaioken ext run: ${error instanceof Error ? error.message : String(error)}\n`);
		return 1;
	}
}

function parseArguments(args: string): Record<string, unknown> {
	const text = args.trim();
	if (!text) return {};
	if (text.startsWith("{")) {
		try {
			const parsed = JSON.parse(text) as unknown;
			if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
				return parsed as Record<string, unknown>;
			}
		} catch {
			// Not JSON after all. Fall through and pass it as the input string,
			// which is more useful than refusing a brace in prose.
		}
	}
	return { input: text };
}

async function resolveEntry(id: string): Promise<Installed | null> {
	if (!id) {
		process.stderr.write("kaioken ext: needs an extension id — `kaioken ext list` shows them\n");
		return null;
	}
	const entry = findInstalled(await loadLock(), id);
	if (!entry) {
		process.stderr.write(`kaioken ext: ${id} is not installed\n`);
		return null;
	}
	return entry;
}

function describeState(entry: Installed): string {
	const parts: string[] = [entry.enabled ? "enabled" : "disabled"];
	if (entry.manifest.type === "declarative") parts.push("documents only");
	else parts.push(isTrusted(entry) ? "trusted to run" : "NOT trusted — inert");
	if (entry.manifest.permissions?.length) parts.push(`asks for ${entry.manifest.permissions.join(", ")}`);
	return parts.join(" · ");
}
